[Bugfix][Enhancement] Improve multi-round-qa.py and data_preprocessing.py scripts

1saac-k opened the pull request:
This PR contains bug fixes and improvements for the `multi-round-qa.py` and `data_preprocessing.py` scripts, which are crucial for benchmarking multi-turn conversations.

The main changes include:
- data_preprocessing.py: excluding invalid traces
- multi-round-qa.py: adding a --dry-run option to quickly test if traces can be processed completely

I've carefully divided the work into individual commits, so if there's an unwanted change, please let me know and I'll drop it.

1saac-k committed:
multi-round-qa: add --dry-run option

This option helps quickly validate trace readability.

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
multi-round-qa: add an assert for insufficient trace data

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
multi-round-qa: fix argument name for init_logger

Using the --verbose option causes an error.

In the init_logger function, `level` should be changed to `log_level`,
or omit it as in this patch.

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
data_preprocessing: add --model option

This option allows you to specify a different tokenizer. The default
value is still `mistralai/Mistral-7B-Instruct-v0.2`.

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
data_preprocessing: adjust estimate_num_tokens location

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
data_preprocessing: exclude invalid conversations

Although there has been some existing logic to exclude invalid
conversations, it is still insufficient.

This patch excludes the following invalid conversations:

- The number of conversations is not even
- The conversation starts with GPT or system
- The conversation ends with human
- GPT or human appears consecutively
- The conversation is too long (may exceed the maximum model length)
- Unknown "from" values

Experimental results using `ShareGPT_V3_unfiltered_cleaned_split.json`
are as follows:

  $ python data_preprocessing.py
  Number of IDs: 94145
  Excluded number of IDs: 58746
  Data to process: 35399

A significant number of traces are excluded.

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
data_preprocessing: use tqdm in loops

The script takes several minutes to run, so it's better to have a
progress bar.

  $ python data_preprocessing.py
  Loading trace file..
  Number of IDs: 94145
  Veryfing trace..
  100%|█████████████████████████| 94145/94145 [01:28<00:00, 1069.70it/s]
  Excluded Number of IDs: 58746
  Data to Process: 35399
  100%|██████████████████████████| 35399/35399 [01:06<00:00, 531.55it/s]

`tqdm` is already a dependency of `vLLM` and `LMCache`, so there's no
additional installation required.

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k committed:
data_preprocessing: add --trace option

This option allows you to specify a trace file. The default value is
still `ShareGPT_V3_unfiltered_cleaned_split.json`.

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k commented:
Added one more minor fix.

Shaoting-Feng commented on benchmarks/multi-round-qa/multi-round-qa.py:
```diff
@@ -521,10 +524,9 @@ def summary(self, start_time: float, end_time: float) -> pd.DataFrame:
 
         df = UserSessionManager.ProcessSummary(
-            df, start_time, end_time, pending_queries, qps
+            df, start_time, end_time, pending_queries
```
The `qps` here is to match the process summary to the various qps values in the run script. It is best to rename the `total_requests / total_time` to something like `real_qps` to avoid overwrite.

1saac-k replied:
Previously, the actual(real) queries per second (`_qps`) were not being printed. This led me to believe that someone had previously attempted to replace the output of `qps` with `_qps`, but failed to do so by mistake.

I will make the following changes:

- Rename actual(real) qps (`_qps`) to `actual_qps`
- Rename config qps (`qps`) to `config_qps`
- Print both `actual_qps` and `config_qps`

1saac-k replied:
After rework:

```
==================== Performance summary ======================
  Config QPS: 10.0000 reqs/s

  Actual QPS: 10.3334 reqs/s

  Processing speed: 10.3334 reqs/s

.. snip ..
```

Shaoting-Feng commented on benchmarks/multi-round-qa/data_preprocessing.py:
```diff
@@ -5,49 +5,124 @@
+data = [d for d in tqdm(data) if not invalid_conversations(d["conversations"])]
+excldued_ids = len(data)
+num_of_ids -= excldued_ids
+print(f"Excluded number of IDs: {excldued_ids}")
```
typo

1saac-k replied:
I'll fix it. Thanks!

Shaoting-Feng reviewed (approved):
LGTM with minor comments

1saac-k committed:
multi-round-qa: fix QPS print statement to print actual QPS value

The print statement outputs qps (configured value) but not _qps (actual
value):

  print(f"  \033[33mQPS: \033[32m{qps:.4f} reqs/s\033[0m\n")

Fix the print statement to output both configured value and actual value.

Changes:

- Rename config qps (`qps`) to `config_qps`
- Rename actual qps (`_qps`) to `actual_qps`
- Print both `actual_qps` and `config_qps`

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k commented:
Hi @ApostaC, It looks like you wrote most of the multi-round-qa script. Would you mind reviewing it when you have a chance? No hurry!

ApostaC commented:
Sure, I will take a look!

ApostaC reviewed (approved):
The multi-round-qa part looks good to me!

@kyet Will there be any changes for people to use `data_preprocessing.py`? If so, can you add it to the README in this folder? Thanks!

1saac-k committed:
doc: update README.md for multi-round-qa and data_preprocessing

Signed-off-by: Kwangjin Ko <kwangjin.ko@sk.com>

1saac-k commented:
@ApostaC  I didn't expect such a quick response! Thank you so much for reviewing at _rocket_ speed—I really appreciate it!

> @kyet Will there be any changes for people to use `data_preprocessing.py`? If so, can you add it to the README in this folder? Thanks!

I've updated the `README.md` :)
